crates/bevy_render/src/picking.rs

```diff
@@ -0,0 +1,434 @@
+            );
+    }
+}
+
+/// Gpu picking uses a double buffer technique, this index is used to know which buffer should be used this frame
+#[derive(Resource)]
+pub struct CurrentBufferIndex(usize);
+fn increment_index(mut curr_index: ResMut<CurrentBufferIndex>) {
+    curr_index.0 = (curr_index.0 + 1) % BUFFER_COUNT;
+}
+
+/// Marker component to indicate that a mesh should be available for gpu picking
+#[derive(Component, ExtractComponent, Clone)]
+pub struct GpuPickingMesh;
+
+/// This component is used to indicate if a camera should support gpu picking.
+/// Any mesh with the [`GpuPickingMesh`] component that is visible from this camera
+/// will be pickable.
+#[derive(Component)]
+pub struct GpuPickingCamera {
+    /// Used to send the required data between the main world and render world
+    data_channel: (Sender<GpuPickingData>, Receiver<GpuPickingData>),
+    /// The latest picking data received
+    data: GpuPickingData,
+    /// Used to determine if the buffer is mapped
+    /// This is only used in the render world, but this is here to avoid needlessly creating a new channel every frame
+    map_status_channel: (Sender<()>, Receiver<()>),
+}
+
+impl Default for GpuPickingCamera {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl GpuPickingCamera {
+    pub fn new() -> Self {
+        Self {
+            data_channel: async_channel::bounded(1),
+            data: GpuPickingData::default(),
+            map_status_channel: async_channel::bounded(1),
+        }
+    }
+
+    /// Get the entity at the given position.
+    /// If there is no entity, returns `None`.
+    pub fn get_entity(&self, pos: UVec2) -> Option<Entity> {
+        // We know the position, but in order to find the true position of the bytes
+        // we're interested in, we have to know how wide a single row in the GPU written buffer is.
+        // Due to alignment requirements this may be wider than the physical camera size because
+        // of padding.
+        let pixel_size = MESH_ID_TEXTURE_FORMAT.pixel_size();
+        let start =
+            (pos.y as usize * self.data.padded_bytes_per_row) + (pos.x as usize * pixel_size);
+        let end = start + pixel_size;
+        if end > self.data.mesh_id_buffer.len() {
+            return None;
+        }
+
+        // TODO This is currently a constant, but could be user configurable
+        let entity = match MESH_ID_TEXTURE_FORMAT {
+            TextureFormat::R16Uint => {
+                let index =
+                    u16::from_ne_bytes(self.data.mesh_id_buffer[start..end].try_into().ok()?);
+                self.data.visible_mesh_entities[index as usize]
+            }
+            TextureFormat::R32Uint => {
+                let index =
+                    u32::from_ne_bytes(self.data.mesh_id_buffer[start..end].try_into().ok()?);
+                self.data.visible_mesh_entities[index as usize]
+            }
+            _ => panic!("Unsupported mesh id texture format"),
+        };
```

Comment:
Nit: Does this or something similar works?
```suggestion
        let texture_bytes = &self.data.mesh_id_buffer[start..end];
        let index = match MESH_ID_TEXTURE_FORMAT {
            TextureFormat::R16Uint => {
                u16::from_ne_bytes(texture_bytes.try_into().ok()?) as usize
            }
            TextureFormat::R32Uint => {
                u32::from_ne_bytes(texture_bytes.try_into().ok()?) as usize
            }
            _ => panic!("Unsupported mesh id texture format"),
        };
        let entity = self.data.visible_mesh_entities[index];
```